### DISCUSSION TITLE
Portfolio Resampling

### DISCUSSION THREAD

**Discussion by sorensenj50** [skfolio/skfolio#5] [Ideas]
Hi everyone, 

I just found out about this package on Twitter. This looks like a really cool project. I'm a quant researcher for a Sovereign Wealth Fund and I'm in the early stages of my own quant python package to help support the community. 

Something that our fund has actually used in portfolio optimization is called portfolio resampling. Not sure if you guys have heard about it. There are a couple of papers out there that introduced it a while back. It is basically a non parametric way to regularize the data by repeatedly adding iid guassian noise to the expected return as well as wishart random variables to the covariance matrix (this keeps it positive semi definite). What you do is repeatedly add this layer of noise to the data and refit the optimizer and then average / aggregate the results to get a resampled frontier. This has the effect of de-regularizing the weights (that is, encouraging diversification) because the expected returns aren't certain. The key parameter to control is the standard deviation of the noise--more noise, more diversification but less certainty on the expected returns, etc. 

If I remember correctly there might be patent around this idea, but it could fit into your uncertaintly set module along with bootstrapping. 

I can look into contributing to this package to build out that functionality if the legal stuff checks out.

**Comment by HugoDelatte** [comment]
Thank you for the suggestion, @sorensenj50. I'm also including @CarloNicolini, who implemented similar algorithms. To provide more context and collaboratively draft the best approach:

Similar to our implementation of StackingOptimization, we should build upon the scikit-learn [BaggingRegressor](https://github.com/scikit-learn/scikit-learn/blob/3f89022fa/sklearn/ensemble/_bagging.py#L993) and extend it, as I'll explain below.

The scikit-learn algorithm encompasses the following:
* Pasting: samples (i.e., observations) are drawn without replacement.
* Bagging: samples are drawn with replacement.
* Random Subspaces: features (i.e., assets) are randomly drawn.
* Random Patches: both samples and features are randomly drawn.

The indices of both samples and features are randomly drawn using [this function](https://github.com/scikit-learn/scikit-learn/blob/3f89022fa04d293152f1d32fbc2a5bdaaf2df364/sklearn/ensemble/_bagging.py#L34).

For portfolio optimization, we need to introduce the following:
* We should incorporate another sampling methodology for observations due to the time correlation of financial series (e.g., block bootstrap).
* As you mentioned, we may consider adding random noise to the observations. It's still unclear whether this should reside outside the BaggingOptimization class and have its own implementation.

**Reply by CarloNicolini** [reply]
@HugoDelatte  in my opinion @sorensenj50 is referring to something similar to what I had done in scikit-portfolio/ensemble called Michaud resampled  efficient frontier. 
[Code here](https://github.com/scikit-portfolio/scikit-portfolio/blob/main/skportfolio/ensemble/_ensemble.py#L43)

Instead of "overfitting" the optimization on a single pair of expected returns/expected risk, one instead normally perturbates the expected returns by `np.random.multivariate_normal` having the `mean` equal to observed expected returns and the `cov` equal to the  sample covariance.
Then one optimizes multiple instances of that and takes the average solution.
This idea is at the basis of the [Michaud and Michaud paper](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2658657) that is criticized by many for not having a strong theoretical fundament.
Also [read here for a deep discussion](https://quant.stackexchange.com/questions/4228/michauds-resampled-efficient-frontier-out-of-sample-simulation-testing) 
I've been reading that the Attilio Meucci Bayesian Robust Allocation solves a lot of the theoretical problems behind this but at the cost of more parameters and algorithmic complication.

BTW I've never heard of also resampling the covariance matrix at the same time through Wishart distribution.
I am not sure how could one sample both at the same time.

**Reply by sorensenj50** [reply]
@CarloNicolini yes I think we're on the same page. I'm glad to see that you have already implemented this feature (or something close to it), at least for expected returns. I am not up to date on the academic discussion but I think this approach has a lot of advantages. The expected returns average out to the initial assumptions, but the uncertainty about the sample estimates are reflected in the model. For instance, if two assets have E[r] of 5% and 5.1%, respectively, portfolio resampling is going to yield a more diversified set of solutions than naive Markowitz. 

The idea of using a Wishart distribution to sample covariance matrices comes from MPI, a modeling platform that our fund used to use more heavily for portfolio optimization. That might be the part that is patented--not sure. Also not sure if this approach has been explored academically. But basically it draws random covariance matrices from a Wishart distribution with scale equal to the original assumed covariance matrix with a parameterized degree of flexibility in terms of the number of observations of the Wishart distribution generative process. 

To your point about perturbing both the expected returns and the covariance matrix, in my modeling I have treated the expected return perturbations as gaussian independent noise with no cross asset correlation. The covariances are then shocked seperately using the wishart method. I haven't compared this method to the method you mention, which is slightly different. Not sure which one is better, but this one is probably more random as the perturbations are completely independent from each other. 

In my mind most of the benefit of portfolio resampling comes from perturbing the expected returns but I don't have any data to back this up. I'm not sure if its worth doing legal work to maybe implement the Wishart covariance piece if the only supporting research was done privately and is probably intellectual property. 

We could think about adding an option in the code you already have to make the perturbations independent vs using the provided covariance matrix, but that would be the only change needed IMO.

**Comment by HugoDelatte** [comment]
Based on your answers, I've started a deep dive into the literature.
There is a vast array  algorithms for portfolio optimization using resampling techniques. 

@CarloNicolini, I noticed that you also implemented subset resampling and robust Bayesian methods which is great. I'll compile what I find in the literature, and then we can discuss the best approach for the API.

**Reply by sorensenj50** [reply]
Sounds good, I'm interested to hear what you find. IMO the more approaches we can implement the better (as long as the API is clear and the users are informed about the pros/cons of said method with academic sources as appropriate) but I'll leave architecture questions up to you guys. 

In the meantime I'll familiarize myself with the codebase and am open to contributing in other ways. I should have some time I can allocate to this over the coming months.

**Reply by CarloNicolini** [reply]
Yes, @HugoDelatte. My implementation of Meucci's Bayesian Robust Allocation is based on the Matlab code provided by Attilio Meucci himself but for Python and with lot of refactoring. The idea is there though. I believe that the robustness parameters set to 0.1 in my default implementation should be selected by means of hyperparameters search with a proper cross-validation approach. Honestly I find it hard to guess values just "based on investor experience".

@sorensenj50 skfolio/scikit-portfolio are thought to have a clear API design and a strong software engineering approach to make it possible to use different methods under the same umbrella.
With @HugoDelatte we have the goal to unify the two libraries to make them the sklearn of portfolio optimization.
This requires thinking deeply around the library design, but I am sure that all methods will fit under this fit/predict estimator-based approach.

**Reply by CarloNicolini** [reply]
I am almost ready with a possible implementation of SubsetSelection based on the sklearn BaggingRegressor.
However I would be grateful to discuss it with you before proposing a pull request @HugoDelatte.

**Reply by HugoDelatte** [reply]
@CarloNicolini  that's amazing, absolutely let's discuss it

**Comment by antonvorobets** [comment]
It is an interesting application of resampling that you use @sorensenj50. Check out this article about resampled portfolio optimization more broadly: https://ssrn.com/abstract=4709317

The article addresses some of the resampling issues mentioned in this discussion and introduces a method that has the potential to outperform equal weighted averaging of the sample portfolios.

Since the Exposure Stacking method operates on portfolio exposures directly, it can be applied to any risk measure and uncertainty in full generality beyond means and covariances.

I will be very happy to assist with adding this functionality. I believe it should be possible to incorporate it nicely into the existing framework.
